### PR TITLE
Nsmd healing

### PR DESCRIPTION
**Pull request by haiodo**
Implement NSMD die Healing for Remote/Local cases.

## Description

* Remote NSMD die handling.
* Source NSMD die handling.
* Local NSE/Client registry for NSMD recovery.
* Recover active connections from Dataplane.

## Motivation and Context
Heal is essential part of NSM, we need to be able to recover from most of die situations.

## How Has This Been Tested?

- [x] Covered by existing integration testing
- [x] Added integration testing to cover
- [x] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: docker/debug/Dockerfile.debug [hunk 1]
```diff
@@ -14,6 +14,12 @@ RUN go get -u golang.org/x/tools/cmd/stringer
 
 RUN curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
 
+ENV GO111MODULE=on
+RUN mkdir /root/networkservicemesh
+ADD ["go.mod","/root/networkservicemesh"]
+WORKDIR /root/networkservicemesh/
+RUN go mod download
+
```
Comment (nickolaev) [inline docker/debug/Dockerfile.debug]: This change probably should belong to another PR.

File: scripts/debug_env.sh [hunk 2]
```diff
@@ -5,14 +5,4 @@ echo "          Welcome to NetworkServiceMesh Dev/Debug environment
...
-echo "Call dep ensure"
-dep ensure
-
-echo "Install openapi-gen"
-go install ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
-echo "Install deepcopy-gen"
-go install ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
-echo "Run generators"
-go generate ./...
+export PS1='DevEnv: \[\e[0;32m\]\u\[\e[m\] \[\e[1;34m\]\w\[\e[m\] \[\e[1;32m\]\$\[\e[m\] \[\e[1;37m\]'
```
Comment (nickolaev) [inline scripts/debug_env.sh]: This is probably for another PR too.

File: test/kube_testing/k8s.go [hunk 3]
```diff
@@ -267,11 +267,10 @@ func NewK8s() (*K8s, error) {
 	return &client, nil
 }
 
-func (o *K8s) deletePods(pods ...*v1.Pod) error {
+func (o *K8s) deletePods(timeout int64, pods ...*v1.Pod) error {
```
Comment (nickolaev) [inline test/kube_testing/k8s.go]: Is this needed for NSMD healing tests? If not strictly needed, I'd suggest to extract ot another PR.

File: test/kube_testing/k8s.go [hunk 4]
```diff
@@ -329,7 +328,7 @@ func (o *K8s) CleanupCRDs() {
 
 func (l *K8s) Cleanup() {
 	for _, result := range l.pods {
-		err := l.deletePods(result)
+		err := l.deletePods(0, result)
```
Comment (nickolaev) [inline test/kube_testing/k8s.go]: What is timeout 0 here?

File: test/kube_testing/pods/nsmd.go [hunk 5]
```diff
@@ -19,15 +24,25 @@ func newDevMount() v1.VolumeMount {
 	}
 }
 
-func NSMDPod(name string, node *v1.Node) *v1.Pod {
-	return createdNSMDPod(name, node, nil, nil)
+func newDevSrcMount() v1.VolumeMount {
+	return v1.VolumeMount{
+		Name:      "src",
+		MountPath: "/go/src",
```
Comment (nickolaev) [inline test/kube_testing/pods/nsmd.go]: Please extract this path to a constant. Should we consider this an ENV variable?

File: test/kube_testing/pods/nsmd.go [hunk 6]
```diff
@@ -94,10 +114,34 @@ func createdNSMDPod(name string, node *v1.Node, liveness, readiness *v1.Probe) *
...
 	if node != nil {
 		pod.Spec.NodeSelector = map[string]string{
 			"kubernetes.io/hostname": node.Labels["kubernetes.io/hostname"],
 		}
 	}
+
+	if nsmdDevImage != "" {
+		pod.Spec.Containers[1].VolumeMounts = append(pod.Spec.Containers[1].VolumeMounts, newDevSrcMount())
+		pod.Spec.Containers[1].Command = []string{ "bash" }
+		pod.Spec.Containers[1].Args = []string{ "/go/src/github.com/networkservicemesh/networkservicemesh/scripts/debug.sh", "nsmd" }
```
Comment (nickolaev) [inline test/kube_testing/pods/nsmd.go]: Constant or ENV variable please.

File: test/kube_testing/pods/nsmd.go [hunk 7]
```diff
@@ -94,10 +114,34 @@ func createdNSMDPod(name string, node *v1.Node, liveness, readiness *v1.Probe) *
...
+		pod.Spec.Volumes = append(pod.Spec.Volumes, v1.Volume{
+			Name: "src",
+			VolumeSource: v1.VolumeSource{
+				HostPath: &v1.HostPathVolumeSource{
+					Type: ht,
+					Path: "/go/src",
```
Comment (nickolaev) [inline test/kube_testing/pods/nsmd.go]: Const or ENV variable please.

File: test/kube_testing/k8s.go (hunk 4 above)
Comment (haiodo) [reply test/kube_testing/k8s.go]: Will update parameter name, it is gracefull timeout, to not wait for pod termination logic.

File: test/kube_testing/pods/nsmd.go (hunk 5 above)
Comment (haiodo) [reply test/kube_testing/pods/nsmd.go]: This is in sync with debug.sh, I will extract this changes to separate PR.

File: scripts/debug_env.sh (hunk 2 above)
Comment (haiodo) [reply scripts/debug_env.sh]: ok

File: docker/debug/Dockerfile.debug (hunk 1 above)
Comment (haiodo) [reply docker/debug/Dockerfile.debug]: ok

File: test/kube_testing/pods/nsmd.go (hunk 6 above)
Comment (haiodo) [reply test/kube_testing/pods/nsmd.go]: will move to constant

File: test/integration/basic_nsmd_deploy_test.go [hunk 8]
```diff
@@ -47,20 +47,20 @@ func testNSMDDdataplaneDeploy(t *testing.T, nsmdPodFactory func(string, *v1.Node
 		return
 	}
 
-	var pods []*v1.Pod
+	var podsVal []*v1.Pod
 
 	for i, node := range nodes {
 		nsmdPodName := fmt.Sprintf("nsmd-%d", i+1)
-		dataPlanePodName := fmt.Sprintf("nsmd-dataplane-%d", i+1)
+		dataPlanePodName := fmt.Sprintf("nsmd-dataplane%d", i+1)
```
Comment (nickolaev) [inline test/integration/basic_nsmd_deploy_test.go]: No. Please let's keep unified naming, We had #817 treating this across all tests.

Overall the changes in this file shoudl nto be part of this PR.

File: test/integration/basic_nsmdp_test.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+	k8s, err := kube_testing.NewK8s()
+	defer k8s.Cleanup()
+
+	Expect(err).To(BeNil())
+
+	k8s.Prepare("nsmd", "icmp")
+
+	nodes := nsmd_test_utils.SetupNodesConfig(k8s, 1, defaultTimeout, []*pods.NSMDPodConfig{})
+	icmpPod := nsmd_test_utils.DeployICMP(k8s, nodes[0].Node, "icmp-responder-nse1", defaultTimeout)
```
Comment (nickolaev) [inline test/integration/basic_nsmdp_test.go]: Please use `icmp-responder-nse-1` with a dash before th number. We had #817 treating this across all tests.

File: test/integration/basic_nsmdp_test.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+	nsmdName := nodes[0].Nsmd.Name
+	k8s.DeletePods(nodes[0].Nsmd)
+	k8s.DeletePods(icmpPod)
+
+	nodes[0].Nsmd = k8s.CreatePod(pods.NSMDPodWithConfig(nsmdName, nodes[0].Node, &pods.NSMDPodConfig{})) // Recovery NSEs
+	icmpPod = nsmd_test_utils.DeployICMP(k8s, nodes[0].Node, "icmp-responder-nse2", defaultTimeout)
```
Comment (nickolaev) [inline test/integration/basic_nsmdp_test.go]: `icmp-responder-nse-2`

File: test/integration/basic_nsmdp_test.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+	k8s, err := kube_testing.NewK8s()
+	defer k8s.Cleanup()
+
+	Expect(err).To(BeNil())
+
+	k8s.Prepare("nsmd", "icmp")
+
+	nodes := nsmd_test_utils.SetupNodesConfig(k8s, 1, defaultTimeout, []*pods.NSMDPodConfig{})
+	icmpPod := nsmd_test_utils.DeployICMP(k8s, nodes[0].Node, "icmp-responder-nse1", defaultTimeout)
```
Comment (nickolaev) [inline test/integration/basic_nsmdp_test.go]: `icmp-responder-nse-1`

File: test/integration/basic_nsmdp_test.go [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+	nsmdName := nodes[0].Nsmd.Name
+	k8s.DeletePods(nodes[0].Nsmd)
+	k8s.DeletePods(icmpPod)
+
+	nodes[0].Nsmd = k8s.CreatePod(pods.NSMDPodWithConfig(nsmdName, nodes[0].Node, &pods.NSMDPodConfig{})) // Recovery NSEs
+	icmpPod = nsmd_test_utils.DeployICMP(k8s, nodes[0].Node, "icmp-responder-nse2", defaultTimeout)
```
Comment (nickolaev) [inline test/integration/basic_nsmdp_test.go]: `icmp-responder-nse-2`

File: test/integration/recover_nse_nsm_heal_test.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+
+	s1 := time.Now()
+	k8s.Prepare("nsmd", "nsc", "nsmd-dataplane", "icmp-responder-nse", "jaeger")
+	logrus.Printf("Cleanup done: %v", time.Since(s1))
+
+	// Deploy open tracing to see what happening.
+	nodes_setup := nsmd_test_utils.SetupNodes(k8s, 2, defaultTimeout )
+
+	// Run ICMP on latest node
+	icmpPod := nsmd_test_utils.DeployICMP(k8s, nodes_setup[1].Node, "icmp-responder-nse1", defaultTimeout)
```
Comment (nickolaev) [inline test/integration/recover_nse_nsm_heal_test.go]: icmp-responder-nse-1

File: test/integration/recover_nse_nsm_heal_test.go [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+	nscPodNode := nsmd_test_utils.DeployNSC(k8s, nodes_setup[0].Node, "nsc1", defaultTimeout, false)
```
Comment (nickolaev) [inline test/integration/recover_nse_nsm_heal_test.go]: nsc-1

File: test/integration/recover_nse_nsm_heal_test.go [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+	nsmdName := fmt.Sprintf("nsmd-worker-recovered-%d", 1)
+
+	logrus.Infof("Starting recovered NSMD...")
+	startTime := time.Now()
+	nodes_setup[1].Nsmd = k8s.CreatePod(pods.NSMDPod(nsmdName, nodes_setup[1].Node)) // Recovery NSEs
+	logrus.Printf("Started new NSMD: %v on node %s", time.Since(startTime), nodes_setup[1].Node.Name)
+
+	failures = InterceptGomegaFailures(func() {
+		// Restore ICMP responder pod.
+		icmpPod = nsmd_test_utils.DeployICMP(k8s, nodes_setup[1].Node, "icmp-responder-nse2", defaultTimeout)
```
Comment (nickolaev) [inline test/integration/recover_nse_nsm_heal_test.go]: icmp-responder-nse-2

File: test/integration/recover_nse_nsm_heal_test.go [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+
+	s1 := time.Now()
+	k8s.Prepare("nsmd", "nsc", "nsmd-dataplane", "icmp-responder-nse", "jaeger")
+	logrus.Printf("Cleanup done: %v", time.Since(s1))
+
+	// Deploy open tracing to see what happening.
+	nodes_setup := nsmd_test_utils.SetupNodes(k8s, 2, defaultTimeout)
+
+	// Run ICMP on latest node
+	icmpPod := nsmd_test_utils.DeployICMP(k8s, nodes_setup[1].Node, "icmp-responder-nse1", defaultTimeout)
```
Comment (nickolaev) [inline test/integration/recover_nse_nsm_heal_test.go]: icmp-responder-nse-1

File: test/integration/recover_nse_nsm_heal_test.go [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+	Expect(icmpPod).ToNot(BeNil())
+
+	nscPodNode := nsmd_test_utils.DeployNSC(k8s, nodes_setup[0].Node, "nsc1", defaultTimeout, false)
```
Comment (nickolaev) [inline test/integration/recover_nse_nsm_heal_test.go]: nsc-1

File: test/kube_testing/k8s.go [hunk 18]
```diff
@@ -144,8 +145,24 @@ func blockUntilPodReady(client kubernetes.Interface, timeout time.Duration, sour
 			break
 		}
 
-		if time.Since(st) > timeout/2 {
+		if time.Since(st) > timeout/2 && !infoPrinted {
 			logrus.Infof("Pod deploy half time passed: %v", pod)
+
+			cmd := exec.Command("kubectl", "get", "pods", "-o", "wide")
```
Comment (nickolaev) [inline test/kube_testing/k8s.go]: Is there a better way? We deliberately left the `shell `script` world.

**Comment by nickolaev** [comment]
@edwarnicke @fkautz we really need to push for smaller PRs. I am sure every complex functionality can be broken down to smaller changes. Reviewing a 45 file PR is just impossible, a source of potential errors and with a high chance of introducing unwanted changes.

**Comment by haiodo** [comment]
Planed to split today, just make it finally working stable.

**Comment by fkautz** [comment]
I'm quite worried about the size of this PR as well. Is it possible to break this up into smaller iterative changes?

**Comment by fkautz** [comment]
Ah you answered and I had a stale window. Thanks @haiodo.

File: test/kube_testing/k8s.go (hunk 18 above)
Comment (haiodo) [reply test/kube_testing/k8s.go]: Removed, I've not found a better way at the moment. Kubernetes API now show all messages some times, not sure why.

**Comment by haiodo** [comment]
Found few more untested cases of local NSM die. Adding tests.

**Comment by haiodo** [comment]
Last failure because, during rebase loose few lines. Do not know how :-(
